klippy/extras/filament_motion_sensor.py

```diff
@@ -0,0 +1,82 @@
+# Filament Motion Sensor Module
+#
+# Copyright (C) 2021 Joshua Wherrett <thejoshw.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import logging
+from . import filament_switch_sensor
+
+CHECK_RUNOUT_TIMEOUT = .250
+
+class MotionHelper:
+    def __init__(self, config):
+        # Read config
+        self.printer = config.get_printer()
+        self.extruder_name = config.get("extruder", None)
+        self.detection_length = config.getfloat(
+                'detection_length', 7., above=0.)
+        # Get printer objects
+        self.reactor = self.printer.get_reactor()
+        self.runout_helper = filament_switch_sensor.RunoutHelper(config)
+        self.extruder = None
+        self.mcu = None
+        # Initialise internal state
+        self.filament_runout_pos = None
+        # Register commands and event handlers
+        self.printer.register_event_handler("klippy:ready",
+                self._handle_ready)
+        self.printer.register_event_handler("idle_timeout:printing",
+                self._handle_printing)
+        self.printer.register_event_handler("idle_timeout:ready",
+                self._handle_not_printing)
+        self.printer.register_event_handler("idle_timeout:idle",
+                self._handle_not_printing)
+    def _update_filament_runout_pos(self, eventtime=None):
+        if eventtime == None:
+            eventtime = self.reactor.monotonic()
+        self.filament_runout_pos = (
+                self._get_extruder_pos(eventtime) +
+                self.detection_length)
+    def _handle_ready(self):
+        self.extruder = self.printer.lookup_object(self.extruder_name)
+        self.mcu = self.extruder.stepper.get_mcu()
+        self._update_filament_runout_pos()
+        self._extruder_pos_update_timer = self.reactor.register_timer(
+                self._extruder_pos_update_event)
+    def _handle_printing(self, print_time):
+        self.reactor.update_timer(self._extruder_pos_update_timer,
+                self.reactor.NOW)
+    def _handle_not_printing(self, print_time):
+        self.reactor.update_timer(self._extruder_pos_update_timer,
+                self.reactor.NEVER)
+    def _get_extruder_pos(self, eventtime=None):
+        if eventtime == None:
```

Comment:
Very minor, but it's preferable to compare with None using `eventtime is None` as classes in Python can override the `==` operator which could confuse things.